### DISCUSSION TITLE
Methods or node-changes?

### DISCUSSION THREAD

**Discussion by om327** [FreeOpcUa/opcua-asyncio#798] [General]
Hello, quick question about best practises - i have a system that comprises of several pumps and valves. 

Pumps tend to have consistent values and properties you can model : Flow Rate, Set Point, state (ON/OFF) and usually have a single method to start and stop the pump. 

When designing a server is it best practise to only have methods triger value changes or for something like updating the set point would it be better to monitor a value change in the server and then this triggers the function to update the set point rather than a new method?

**Comment by AndreasHeine** [comment]
for valuechanges -> write
for start/stop -> method (because you can react with Bad_NotExecutable status if the pump has a failure)

nice to know:
there is a OPC UA CompanionSpec. which already have models for pumps in OPC UA!
https://reference.opcfoundation.org/Pumps/docs/

**Comment by om327** [comment]
(because you can react with Bad_NotExecutable status if the pump has a failure) - how do you mean?

**Reply by AndreasHeine** [reply]
if a client calls a OPC UA Method in a OPC UA Server the Server can react with StatusCode "Good" and the OutArguments will be send to the client or the Server can responde with a Statuscode "Bad_..."

References:
https://reference.opcfoundation.org/Core/docs/Part4/5.11.1/
see: "Call Service Result Codes" and "Call Operation Level Result Codes"
https://reference.opcfoundation.org/v105/Core/docs/Part4/5.11.2/

**Reply by AndreasHeine** [reply]
in simple words: 
if the pump has a interlock to start only if some condition is met
the server can tell the client if he calls the method that it is not possible!
